### PR TITLE
Multi variable scan

### PR DESCRIPTION
**Pull request by kraemeru**
Added a new variable "scanParallelTo" to the scanHelper class and .scan files to allow the change of multiple variables of the same or separate producers in a single scan step at the same time.

If this variable is not specified in the .scan file it simply behaves like the old scan to ensure it is backwards compatible.

### CODE REVIEW

File: gui/src/scanHelper.cc [hunk 1]
```diff
@@ -73,42 +92,103 @@ bool Scan::setupScan(std::string globalConfFile, std::string scanConfFile) {
     return true;
 }
 
-void Scan::createConfigs(unsigned condition, eudaq::ConfigurationSP conf,std::vector<ScanSection> sec) {
-    if(condition == sec.size())
+//void Scan::createConfigs(unsigned condition, eudaq::ConfigurationSP conf,std::vector<ScanSection> sec) {
+//    if(condition == sec.size())
```
Comment (lhuth) [inline gui/src/scanHelper.cc:96]: Commented code can and should be removed

File: user/example/misc/scan/ExampleScan.scan [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [global]
-repeatScans = 1
+repeatScans = 0
```
Comment (lhuth) [inline user/example/misc/scan/ExampleScan.scan:2]: revert changes that are not relevant for the PR, like this one :)

File: user/example/misc/scan/ExampleScan.scan [hunk 3]
```diff
@@ -16,7 +16,7 @@ name = Producer.my_pd0
 parameter = EX0_PLANE_ID
 
 [4]
-nested = 1
+#nested = 1
```
Comment (lhuth) [inline user/example/misc/scan/ExampleScan.scan:19]: or this :)

File: user/example/misc/example_startrun.sh [hunk 4]
```diff
@@ -5,8 +5,8 @@ $BINPATH/euRun -n Ex0RunControl &
 sleep 1
 $BINPATH/euLog &
 sleep 1
-$BINPATH/euCliMonitor -n Ex0Monitor -t my_mon & 
-$BINPATH/euCliCollector -n Ex0TgDataCollector -t my_dc &
+#$BINPATH/euCliMonitor -n Ex0Monitor -t my_mon & 
+#$BINPATH/euCliCollector -n Ex0TgDataCollector -t my_dc &
```
Comment (lhuth) [inline user/example/misc/example_startrun.sh:9]: Removing this makes the script obsolete. Any reason why this is done?

File: user/example/misc/scan/ExampleScan2.scan [hunk 5]
```diff
@@ -0,0 +1,40 @@
+[global]
+repeatScans = 0
+allowNested = 1 # overrides local nested arguments
+#configPrefix = "path/to/folder/scanned" // optional
+timeBasedScan = 1 # 1 = true :)
+timePerStep = 10 #in seconds for now
+nEventsPerStep   = 200
+
+[0]
+scanParallelTo = 0
```
Comment (lhuth) [inline user/example/misc/scan/ExampleScan2.scan:10]: Why would I like to scan that in parallel to itself? Looks odd to me. What if it is not defined at all?

File: gui/include/scanHelper.hh [hunk 6]
```diff
@@ -126,6 +128,15 @@ private:
      * @param sec Scan section beeing read
      */
     void createConfigs(unsigned condition, eudaq::ConfigurationSP conf, std::vector<ScanSection> sec);
```
Comment (lhuth) [inline gui/include/scanHelper.hh:130]: If I am not blind you have removed the function in the cpp. Why is it still here? :)

**Review by lhuth — CHANGES_REQUESTED** [review]
Hi @kraemeru,

thanks for setting this up. Overall a nice feature. A few more or less picky comments are below.

File: user/example/misc/scan/ExampleScan2.scan (hunk 5 above)
Comment (kraemeru) [reply user/example/misc/scan/ExampleScan2.scan:10]: I am not sure what you mean by "that in parallel to itself" if it is about the fact that it affects the same producer this is simply to show that you can also change two variables of the same producer at the same time.

If scanParallelTo is not defined it simply counts each [X] variable as its own unit to ensure backwards compatibility with old scan files.

File: user/example/misc/example_startrun.sh [hunk 7]
```diff
@@ -5,11 +5,11 @@ $BINPATH/euRun -n Ex0RunControl &
...
-$BINPATH/euCliMonitor -n Ex0Monitor -t my_mon & 
+$BINPATH/euCliMonitor -n Ex0Monitor -t my_mon &
 $BINPATH/euCliCollector -n Ex0TgDataCollector -t my_dc &
 # The following data collectors are provided if you build user/eudet
-#$BINPATH/euCliCollector -n DirectSaveDataCollector -t my_dc &
-#$BINPATH/euCliCollector -n EventIDSyncDataCollector -t my_dc &
-#$BINPATH/euCliCollector -n TriggerIDSyncDataCollector -t my_dc &
+$BINPATH/euCliCollector -n DirectSaveDataCollector -t my_dc &
+$BINPATH/euCliCollector -n EventIDSyncDataCollector -t my_dc &
+$BINPATH/euCliCollector -n TriggerIDSyncDataCollector -t my_dc &
```
Comment (lhuth) [inline user/example/misc/example_startrun.sh:13]: The -t <> option has to provide a unique name.

File: user/example/misc/scan/ExampleScan2.scan (hunk 5 above)
Comment (lhuth) [reply user/example/misc/scan/ExampleScan2.scan:10]: got it

File: gui/include/scanHelper.hh [hunk 8]
```diff
@@ -5,6 +5,7 @@
 #include <string>
 #include <vector>
 #include <cstdlib>
+#include <map>
```
Comment (lhuth) [inline gui/include/scanHelper.hh:8]: at least on my machine `#include <unordered_map>` is required to compile

**Review by lhuth — CHANGES_REQUESTED** [review]

File: user/example/misc/scan/ExampleScan2.scan [hunk 9]
```diff
@@ -0,0 +1,40 @@
+[global]
```
Comment (lhuth) [inline user/example/misc/scan/ExampleScan2.scan:1]: Maybe you can give the multi parameter scan file a more meaningful name?

File: user/example/misc/example_startrun.sh (hunk 7 above)
Comment (kraemeru) [reply user/example/misc/example_startrun.sh:13]: I don't remember doing this change but maybe I simply misremember? But I now called them something else specific  to the type of data collector.

**Review by lhuth — APPROVED** [review]

**Pull request merged**